### PR TITLE
fix: remove  nin and ne  operator usage in the SDK and the sample app

### PR DESCRIPTION
**Pull request by khushal87**
## 🎯 Goal

The `$ nin` and the `$ne` operators cause bad performance on the backend, and they will be deprecated (and eventually removed). We should also start removing them from the SDKs and the sample app and not advocate integrators to use it. 

A few simplifications around the query users filter, and the case where the channel state members are less than 100 are also done in this PR.

Few other fixes in the PR:
- Fix the bug where you would get the authenticated user as well in the users list when you try to mention it.
- Restructure the functions to separate files.

## 🎨 UI Changes

<details>
<summary>iOS</summary>

<table>
    <thead>
        <tr>
            <td>Before</td>
            <td>After</td>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                
            </td>
            <td>
                
            </td>
        </tr>
    </tbody>
</table>
</details>

<details>
<summary>Android</summary>

<table>
    <thead>
        <tr>
            <td>Before</td>
            <td>After</td>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                
            </td>
            <td>
                
            </td>
        </tr>
    </tbody>
</table>
</details>

## ☑️ Checklist

- [ ] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [ ] PR targets the `develop` branch
- [ ] Documentation is updated
- [ ] New code is tested in main example apps, including all possible scenarios
  - [ ] SampleApp iOS and Android
  - [ ] Expo iOS and Android

### CODE REVIEW

**Review by isekovanic — APPROVED** [review]

File: package/src/utils/utils.ts [hunk 1]
```diff
@@ -234,32 +229,30 @@ const queryUsers = async <
...
     );
-    const users: SuggestionUser<StreamChatGenerics>[] = [];
-    response.users.forEach((user) => isUserResponse(user) && users.push(user));
+    const usersWithoutClientUserId = users.filter((user) => user.id !== client.userID);
     if (onReady && users) {
-      onReady(users);
+      onReady(usersWithoutClientUserId);
     }
+  } catch (error) {
+    console.warn('Error querying users:', error);
```
Comment (oliverlaz) [inline package/src/utils/utils.ts:255]: Shouldn't we bubble this error up? How can one recover from an error that can happen here?

File: package/src/utils/utils.ts [hunk 2]
```diff
@@ -168,25 +160,28 @@ const getWatchers = <
...
-        acc[cur.id] = cur;
-      }
+  // make sure we don't list users twice
+  const uniqueUsers = {} as Record<string, UserResponse<StreamChatGenerics>>;
+
+  users.forEach((user) => {
+    if (user && !uniqueUsers[user.id]) {
+      uniqueUsers[user.id] = user;
+    }
+  });
```
Comment (oliverlaz) [inline package/src/utils/utils.ts:182]: Unrelated, but I think you can speed up this function and save one full object iteration (`Object.values()`) if you rewrite it like this:

```ts
const seenUsers = new Set<string>();
const uniqueUsers: User[] = [];
for (const user of users) {
  if (user && !seenUsers.has(user.id)) {
    uniqueUsers.push(user);
    seenUsers.add(user.id);
  }
}
return uniqueUsers;
```

File: examples/SampleApp/src/hooks/usePaginatedUsers.ts [hunk 3]
```diff
@@ -153,33 +150,35 @@ export const usePaginatedUsers = (): PaginatedUsers => {
...
       // Dumb check to avoid duplicates
-      if (query === searchText && results.findIndex((r) => res?.users[0].id === r.id) > -1) {
+      if (
+        query === searchText &&
+        usersWithoutClientUserId.length > 0 &&
+        results.findIndex((r) => usersWithoutClientUserId[0].id === r.id) > -1
+      ) {
         queryInProgress.current = false;
         return;
       }
```
Comment (oliverlaz) [inline examples/SampleApp/src/hooks/usePaginatedUsers.ts:163]: Again unrelated but, this is quite confusing to me. What does this code do?

Comment (khushal87) [reply examples/SampleApp/src/hooks/usePaginatedUsers.ts:163]: We can remove it imo. Seems useless now

File: package/src/utils/utils.ts [hunk 4]
```diff
@@ -2,17 +2,16 @@ import type React from 'react';
 
 import dayjs from 'dayjs';
 import EmojiRegex from 'emoji-regex';
-import type { DebouncedFunc } from 'lodash';
+import { DebouncedFunc } from 'lodash';
```
Comment (oliverlaz) [inline package/src/utils/utils.ts:5]: Let's revert this one. I'm not sure if this change will lead to bundling the whole `lodash` library.

File: package/src/utils/queryMembers.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+      {
+        name: { $autocomplete: query },
+      },
+      {},
+      { limit },
+    )) as ChannelMemberAPIResponse<StreamChatGenerics>;
+
+    const users: SuggestionUser<StreamChatGenerics>[] = [];
+    members
+      .filter((member) => member.user?.id !== client.userID)
```
Comment (khushal87) [inline package/src/utils/queryMembers.ts:94]: This was not present before and is added in this PR. Fixes returning of authenticated user in the suggestions as well.

**Review by oliverlaz — APPROVED** [review]

**Pull request merged**